app/src/main/java/com/google/droidjump/LevelsFragment.java

```diff
@@ -17,25 +17,72 @@
 package com.google.droidjump;
 
 import static androidx.navigation.Navigation.findNavController;
+import static com.google.droidjump.GameConstants.GAME_VIEW_LEVEL_STRING;
+import android.annotation.SuppressLint;
+import android.content.Context;
 import android.os.Bundle;
+import android.support.annotation.Nullable;
 import android.support.v4.app.Fragment;
 import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
-import android.widget.Button;
+import android.widget.GridView;
+import android.widget.ImageButton;
+import java.util.Objects;
 
 /**
  * Displays Levels Screen.
  */
 public class LevelsFragment extends Fragment {
 
+    private MainActivity activity;
+    private int[] levels;
+    private LevelsAdapter adapter;
+
     @Override
-    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
-        View rootView = inflater.inflate(R.layout.levels_screen, container, /* attachToRoot= */ false);
-        Button levelPlayButton = rootView.findViewById(R.id.level_play);
-        levelPlayButton.setOnClickListener(view -> {
-            findNavController(view).navigate(R.id.action_levels_screen_to_game_screen);
+    public void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        activity = Objects.requireNonNull((MainActivity) getActivity());
+        levels = createLevels(activity.getLevelsCount());
+        Context context = getContext();
+        adapter = new LevelsAdapter(Objects.requireNonNull(context), levels,
+                activity.getCurrentLevel());
+    }
+
+    @SuppressLint("SetTextI18n")
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+        View rootView = inflater.inflate(R.layout.levels_screen, container,
+                /* attachToRoot= */ false);
+
+        // Finding gridView and putting levels to it
+        GridView gridView = rootView.findViewById(R.id.levels_grid_view);
+        gridView.setAdapter(adapter);
+
+        // Adding onClick events
+        gridView.setOnItemClickListener((adapterView, view, i, l) -> {
+            if (activity.getCurrentLevel() >= (int) adapter.getItem(i)) {
+                Bundle args = new Bundle();
+                args.putInt(GAME_VIEW_LEVEL_STRING, (int) adapter.getItem(i));
```

Comment:
If I understand correctly, this will wipe the player's progress if they select some other level other than the last. Please store two pieces of data - the last level that player can play, and the currently played level.

Reply:
No, we just pass the level that we are going to play in gameView. If this level equals the currentLevel, then after the player wins, we increase the currentLevel.